ouroboros-consensus/src/consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Honest.hs

```diff
@@ -0,0 +1,650 @@
+import           Prelude hiding (words)
+import qualified System.Random.Stateful as R
+import qualified Test.Ouroboros.Consensus.ChainGenerator.BitVector as BV
+import qualified Test.Ouroboros.Consensus.ChainGenerator.Counting as C
+import           Test.Ouroboros.Consensus.ChainGenerator.Params (Asc,
+                     Delta (Delta), Kcp (Kcp), Len (Len), Scg (Scg), ascVal)
+import qualified Test.Ouroboros.Consensus.ChainGenerator.Slot as S
+import           Test.Ouroboros.Consensus.ChainGenerator.Slot
+                     (E (ActiveSlotE, SlotE), S)
+import qualified Test.Ouroboros.Consensus.ChainGenerator.Some as Some
+
+-----
+
+data HonestRecipe = HonestRecipe !Kcp !Scg !Delta !Len
+  deriving (Eq, Read, Show)
+
+-- | A specification of which density to give to a window in a containing
+-- sequence of slots
+--
+-- * @base@ the type-level name of the containing sequence
+-- * @hon@ the type-level name of the window in which the density must be
+--   ensured (e.g. @Win (Lbl \@HonestLbl)@)
+--
+-- TODO: Rename to CheckedHonestSchemaSpec
+data CheckedHonestRecipe base hon = UnsafeCheckedHonestRecipe {
+    -- | Desired density
+    chrScgDensity :: !(BV.SomeDensityWindow S.NotInverted)
+  , -- | Window in the @base@ containing sequence where the density should be
+    -- ensured
+    chrWin         :: !(C.Contains SlotE base hon)
+  }
+  deriving (Eq, Read, Show)
+
+-- TODO: Rename to SomeCheckedHonestSpec
+data SomeCheckedHonestRecipe =
+    forall base hon.
+    SomeCheckedHonestRecipe
+        !(Proxy base)
+        !(Proxy hon)
+        !(CheckedHonestRecipe base hon)
+
+instance Show SomeCheckedHonestRecipe where
+    showsPrec p (SomeCheckedHonestRecipe base hon recipe) =
+        Some.runShowsPrec p
+      $ Some.showCtor SomeCheckedHonestRecipe "SomeCheckedHonestRecipe"
+            `Some.showArg` base
+            `Some.showArg` hon
+            `Some.showArg` recipe
+
+instance Read SomeCheckedHonestRecipe where
+    readPrec =
+        Some.runReadPrec
+      $ Some.readCtor SomeCheckedHonestRecipe "SomeCheckedHonestRecipe"
+            <*> Some.readArg
+            <*> Some.readArg
+            <*> Some.readArg
+
+data NoSuchHonestChainSchema =
+    -- | 'Scg' must be greater than 'Delta'
+    BadDeltaScg
+  |
+    -- | must have @0 <= 'Kcp' < 'Scg' - 'Delta'@
+    BadKcp
+  |
+    -- | 'Len' must be positive
+    BadLen
+  deriving (Eq, Read, Show)
+
+checkHonestRecipe :: HonestRecipe -> Exn.Except NoSuchHonestChainSchema SomeCheckedHonestRecipe
+checkHonestRecipe recipe = do
+    when (s <= d) $ Exn.throwError BadDeltaScg
+
+    when (l <= 0) $ Exn.throwError BadLen
+
+    when (k < 0 || s - d <= k) $ Exn.throwError BadKcp
+
+    C.withTopWindow (C.Lbl @HonestLbl) l $ \base topWindow -> do
+        C.SomeWindow Proxy slots <- pure topWindow
+
+        pure $ SomeCheckedHonestRecipe base Proxy UnsafeCheckedHonestRecipe {
+            chrScgDensity = BV.SomeDensityWindow (C.Count (k + 1)) (C.Count s)
```

Comment:
Changing the density here to `k` instead of `k+1` breaks the tests, and I'm guessing that it will require fixing the generation of the first stability window of the schema. 

I would need to have the generation simplified before doing this change, or I would need to pair with someone to do it.

Reply:
Solved in db26c4e and the immediately following commits.